pod-operation/src/components/pressure_transducer.rs

```diff
@@ -2,51 +2,101 @@ use ina219::INA219;
 // mA. A scaling factor is needed in order to convert the raw reading to mA and
 // this is not provided in the INA219 library that we are using. Note that this
 // value changes according to the calibration value. The exact formula can be
 // found in the INA219 datasheet.
 const INA219_SCALING_VALUE: f32 = 160.0;
 
-// The pressure transducer outputs a current between 4 mA and 20 mA with 0 PSI
-// and 300 PSI respectively. Assuming a linear interpolation, a 1 mA increase
-// results in a 18.75 PSI increase.
-const REF_CURRENT_LOW: f32 = 4.0;
-const REF_CURRENT_HIGH: f32 = 20.0;
-const REF_PRESSURE_LOW: f32 = 0.0;
-const REF_PRESSURE_HIGH: f32 = 300.0;
+struct Reference {
+	pressure_lo: f32,
+	pressure_span: f32,
+	current_lo: f32,
+	current_span: f32,
+}
+
+impl Reference {
+	fn new(pressure_lo: f32, pressure_hi: f32, current_lo: f32, current_hi: f32) -> Self {
+		Self {
+			pressure_lo,
+			current_lo,
+			pressure_span: pressure_hi - pressure_lo,
+			current_span: current_hi - current_lo,
+		}
+	}
 
-const REF_CURRENT_SPAN: f32 = REF_CURRENT_HIGH - REF_CURRENT_LOW;
-const REF_PRESSURE_SPAN: f32 = REF_PRESSURE_HIGH - REF_PRESSURE_LOW;
+	// The upstream pressure transducer outputs a current between 4 mA and 20 mA
+	// with 0 PSI and 5000 PSI respectively.
+	fn upstream() -> Self {
+		Self::new(0.0, 5000.0, 4.0, 20.0)
+	}
 
-impl PressureTransducer {
-	pub fn new(ina219_addr: u8) -> Self {
-		let device = I2c::new().unwrap();
+	// The downtream pressure transducer outputs a current between 4 mA and 20 mA
+	// with 0 PSI and 300 PSI respectively.
+	fn downstream() -> Self {
+		Self::new(0.0, 300.0, 4.0, 20.0)
+	}
+}
 
-		let mut ina219 = INA219::new(device, ina219_addr);
-		debug!("Initialized I2C and INA219");
+fn init_ina(device_address: u8) -> INA219<I2c> {
+	let device = I2c::new().unwrap();
 
-		ina219.calibrate(INA219_CALIBRATION_VALUE).unwrap();
-		debug!("Calibrating INA219");
+	let mut ina219 = INA219::new(device, device_address);
+	debug!("Initialized I2C and INA219");
 
-		PressureTransducer { ina: ina219 }
+	ina219.calibrate(INA219_CALIBRATION_VALUE).unwrap();
+	debug!("Calibrating INA219");
+
+	ina219
+}
+
+pub struct PressureTransducer {
+	ina: INA219<I2c>,
+	ref_values: Reference,
+}
+
+impl PressureTransducer {
+	// This constructor should be used for INA219s where the address pins are
+	// grounded. That is, the device address is 0x40.
+	pub fn upstream() -> Self {
+		Self {
+			ina: init_ina(INA219_UPSTREAM_ADDRESS),
+			ref_values: Reference::upstream(),
+		}
+	}
+
+	// This constructor should be used for INA219s where the address pin A0 is
+	// jumped. That is, the device address is 0x41.
+	pub fn downstream() -> Self {
+		Self {
+			ina: init_ina(INA219_DOWNSTREAM_ADDRESS),
+			ref_values: Reference::downstream(),
+		}
 	}
 
 	// Read current from the INA219 and apply a scaling factor to translate
 	// the current reading to PSI.
 	pub fn read(&mut self) -> f32 {
```

Comment:
Suggestion: Nitpick but could we rename `read` to `read_pressure` for improved clarity. Otherwise looks good!